File: Algorithms/Tree and Graph/BFS.swift
//
//  BFS.swift
//  Algorithms
//
//  Created by Rahul Ranjan on 3/29/17.
//  Copyright © 2017 Rahul Ranjan. All rights reserved.
//

import Foundation

func breadthFirstSearch(graph: AdjancencyListGraph<String> , source: Vertex<String>) -> [Vertex<String>]? {
    
  var queue = Queue<Vertex<String>>()
  var visited = { () -> [Vertex<String> : Bool] in
    var visited: [Vertex<String>: Bool] = [:]
    graph.adjacencyDict.keys.forEach {
      visited[$0] = false
    }
    return visited
  }()

  var path: [Vertex<String>] = []
  queue.enqueue(source)
  visited[source] = true

  while !queue.isEmpty {
    let topElement = queue.dequeue()
    path.append(topElement!)

    guard let neigbhours = graph.edges(from: topElement!) else {
      return path
    }

    for neigbhour in neigbhours {
      let destinationVertex = neigbhour.destination

      if visited[destinationVertex] == false {
        queue.enqueue(destinationVertex)
        visited[destinationVertex] = true
      }
    }
  }
  return path
}

func breadthFirstSearchTest() {
  let (graph, source) = initialiseAdjacencyListGraph()
  let path = breadthFirstSearch(graph: graph, source: source)

  print(path!)
}

File: Algorithms/Python/Graph/BFS.py
#!/usr/bin/python3

from queue import Queue

def BFS(source, graph):
	path = list()
	v = {k:0 for k in graph.keys()}
	
	# mark source node visited and put it in
	# the queue
	q = Queue()
	v[source] = 1
	q.put(source)
	
	while not q.empty():
		t = q.get()
		path.append(t)

		for n in graph[t]:
			if v[n] == 0:
				v[n] = 1
				q.put(n)
	return path

def testBFS():
	graph = {
		1: [2, 4],
		2: [3, 1, 5],
		3: [2, 5],
		5: [2, 3],
		4: [1]
	}
	
	path = BFS(1, graph)
	print(path)

testBFS()